wit/atomic.wit

```diff
@@ -1,31 +1,32 @@
 /// A keyvalue interface that provides atomic operations.
 /// 
-/// Atomic operations are single, indivisible operations. When a fault causes
-/// an atomic operation to fail, it will appear to the invoker of the atomic
-/// operation that the action either completed successfully or did nothing
-/// at all.
-interface atomic {
-	/// A keyvalue interface that provides atomic operations.
-	use types.{bucket, error, key};
+/// Atomic operations are single, indivisible operations. When a fault causes an atomic operation to
+/// fail, it will appear to the invoker of the atomic operation that the action either completed
+/// successfully or did nothing at all.
+interface atomics {
+  use types.{bucket, error};
+  
+  /// The atomics resource contains atomic operations that can be executed against a store
+  resource atomics {
```

Comment:
I don't think at the moment `wit-parser` can parse `%[method]`.

```bash
error: failed to parse package: wit

Caused by:
    0: identifiers must have characters between '-'s
            --> wit/atomic.wit:33:2
             |
          33 |  %[method]bucket.increment: func(key: string, delta: u64) -> result<u64, error>;
             |  ^
```

But I like this idea of extending resources. This might be something we could work on the upstream CM spec.

Reply:
Ah, you're right; thanks for testing that out!  The syntax is included in WIT.md [here](https://github.com/WebAssembly/component-model/blob/main/design/mvp/WIT.md#item-resource) which is why I thought it might, so maybe this is just a Not Yet Implemented case.  @alexcrichton would allowing `%[method]bucket.increment` to parse as a function name be an easy fix for someone to make?

Reply:
Oh that's intentional actually, there's two issues here:

* One is that wit-parser doesn't support the syntax as-written here, and yeah that'd be an easy fix.
* The second more fundamental issue is that component validation requires all methods for a resource to be defined in the same interface.

Currently the syntax supported is intended to reflect the second requirement by requiring methods in the `resource` block.

Reply:
Same as my other comment, but basically this means we can't do this right now, though we'd love to do it if we can in the future. That correct?

Reply:
Ah, gotcha, thanks!  Well yes, that's probably a more nuanced C-M discussion for later.

In the meantime, even this `%[method]bucket.increment` name is just syntactic sugar (for adding `increment` as a method to the trait/class/prototype of `bucket` in the bindings). But we could also just write:
```wit
interface atomics {
  use store.{bucket};
  increment: func(borrow<bucket>, key:string, delta:u32);
  ...
}
```
and sure you'll have to call `increment(bucket, `key', 1)` instead of `bucket.increment('key', 1)`, but it works just as well, so perhaps that's the right starting point (so that we get the structure of the API right, before worrying about method-vs-function call syntax and [C-M/#296](https://github.com/WebAssembly/component-model/issues/296) properly)).

Reply:
I think one of the changes this PR introduced is to remove the extra `borrow<bucket>` parameter from the function signature, which we have them today, correct me if I am wrong @thomastaylor312 . Are you suggesting that we should reverse this change?

Reply:
Oh, I'd like them to be on handles, it's just that we don't have the WIT functionality yet to make this happen from interfaces other than the one that defined the resource.  So, e.g., in `wasi:keyvalue/store`, we can have:
```wit
interface store {
  resource bucket {
    get: ...
    set: ...
  }
}
```
it's only for `atomics` and `batch` where I'm suggesting that we define them as non-member functions (for lack of the above WIT feature).

The alternative, giving each of `atomics` and `batch` their own resources, seems to add a whole new dimension of dynamism to the interface and more-fundamentally change its structure, so I'm suggesting we don't do that just so we can write `a.foo()` instead of `foo(a)` (and only for `atomics` and `batch`).

Reply:
Yeah, that makes a lot of sense. I've pushed a new commit with those changes

Reply:
Thanks for the consideration, looks good!